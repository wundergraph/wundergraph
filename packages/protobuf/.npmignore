node_modules
wgpb
tsconfig.json
.gitignore
.npmignore
.npmrc
dist/tsconfig.tsbuildinfo
