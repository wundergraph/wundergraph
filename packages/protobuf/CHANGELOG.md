# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.109.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.108.0...@wundergraph/protobuf@0.109.0) (2023-04-24)

### Features

* add support for using proxies in data sources ([#854](https://github.com/wundergraph/wundergraph/issues/854)) ([a472c46](https://github.com/wundergraph/wundergraph/commit/a472c463d6dbcb20b48444fdd4ce9c797c16feb2)) (@fiam)

## [0.108.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.107.0...@wundergraph/protobuf@0.108.0) (2023-04-19)

### Features

* implement internal graphql handler ([#853](https://github.com/wundergraph/wundergraph/issues/853)) ([ad42f5f](https://github.com/wundergraph/wundergraph/commit/ad42f5f9f23a0e6ec968a1c5fefb226f2b46254a)) (@thisisnithin)

## [0.107.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.106.1...@wundergraph/protobuf@0.107.0) (2023-04-12)

### Features

* tui + log handling ([#813](https://github.com/wundergraph/wundergraph/issues/813)) ([6bf7498](https://github.com/wundergraph/wundergraph/commit/6bf74980c69def43e4af8bdfdcc0fb5645a1025b)) (@StarpTech)

## [0.106.1](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.106.0...@wundergraph/protobuf@0.106.1) (2023-03-16)

### Bug Fixes

* custom int/float scalars usage in arguments ([#427](https://github.com/wundergraph/wundergraph/issues/427))Co-authored-by: Alberto GarciÃÅa Hierro <alberto@garciahierro.com> Co-authored-by: Sergiy <818351+devsergiy@users.noreply.github.com> Co-authored-by: spetrunin <neyasut@gmail.com> ([2b424f3](https://github.com/wundergraph/wundergraph/commit/2b424f3daaabea25abb4a23139751ff0ab2adefe)) (@OLingard)

## [0.106.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.105.0...@wundergraph/protobuf@0.106.0) (2023-03-06)

### Features

* add support for declaring a list of public claims ([#706](https://github.com/wundergraph/wundergraph/issues/706)) ([2223d5c](https://github.com/wundergraph/wundergraph/commit/2223d5c02f1eadaad7ecb6ac863bb2f9e34c926c)) (@fiam)
* add support for injecting variables into nested paths ([#678](https://github.com/wundergraph/wundergraph/issues/678)) ([ce0a0a0](https://github.com/wundergraph/wundergraph/commit/ce0a0a09b4e739767b17f1ad391074f6ba597253)) (@fiam)

## [0.105.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.104.0...@wundergraph/protobuf@0.105.0) (2023-02-10)

### Features

* add support for custom claims ([#648](https://github.com/wundergraph/wundergraph/issues/648) ([682b88c](https://github.com/wundergraph/wundergraph/commit/682b88c115512878aa22b57d4ed67c07fd88726b)) (@fiam)

## [0.104.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.103.0...@wundergraph/protobuf@0.104.0) (2023-02-03)

### Features

* add support for uploads from anonymous users ([#632](https://github.com/wundergraph/wundergraph/issues/632)) ([485975b](https://github.com/wundergraph/wundergraph/commit/485975be55a8c1c82a4610431c7514d3c3c2dfe0)) (@fiam)

## [0.103.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.102.0...@wundergraph/protobuf@0.103.0) (2023-01-30)

### Features

* add prisma datasource ([#605](https://github.com/wundergraph/wundergraph/issues/605)) ([2f0cfa5](https://github.com/wundergraph/wundergraph/commit/2f0cfa5e420080772a6552f4a256c7d71906f84a)) (@jensneuse)

## [0.102.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.101.0...@wundergraph/protobuf@0.102.0) (2023-01-20)

### Features

* add upload profiles for S3 providers ([#476](https://github.com/wundergraph/wundergraph/issues/476)) ([6144545](https://github.com/wundergraph/wundergraph/commit/614454539133c7f235aea6aa72ade36059f41c97)) (@fiam)

## [0.101.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.100.0...@wundergraph/protobuf@0.101.0) (2023-01-17)

### Features

* add support for disconnecting from Auth0 ([#525](https://github.com/wundergraph/wundergraph/issues/525)) ([7465fb2](https://github.com/wundergraph/wundergraph/commit/7465fb21a3618924c7dfb59a6a2f94c7d740f0f8)) (@fiam)

## [0.100.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.99.0...@wundergraph/protobuf@0.100.0) (2022-12-06)

### Features

* make detecting invalid operations easier ([#374](https://github.com/wundergraph/wundergraph/issues/374)) ([45639db](https://github.com/wundergraph/wundergraph/commit/45639db0ae3adb8cac4f62b623b04061118f7bf7)) (@fiam)

## [0.99.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.98.0...@wundergraph/protobuf@0.99.0) (2022-11-18)

### Features

* remove app/main from URL structure ([#335](https://github.com/wundergraph/wundergraph/issues/335)) ([e49e585](https://github.com/wundergraph/wundergraph/commit/e49e585528297126b93958105e80b68f1943d781)), closes [#333](https://github.com/wundergraph/wundergraph/issues/333) (@fiam)
* sdk, schema extension fields config ([#336](https://github.com/wundergraph/wundergraph/issues/336)) ([ca09b3c](https://github.com/wundergraph/wundergraph/commit/ca09b3cf2002763b7ea53a7d50f6dce50d08c120)) (@YuriBuerov)

## [0.98.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.97.0...@wundergraph/protobuf@0.98.0) (2022-11-04)

### Features

* eng 640 compose subgraphs without apollo dependencies ([#315](https://github.com/wundergraph/wundergraph/issues/315)) ([628a183](https://github.com/wundergraph/wundergraph/commit/628a18303acb47df5a10118b17a4e88916b2903a)) (@jensneuse)

## [0.97.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.96.0...@wundergraph/protobuf@0.97.0) (2022-10-18)

### Features

* add configurable per-source timeouts ([#232](https://github.com/wundergraph/wundergraph/issues/232)) ([bb3b6bd](https://github.com/wundergraph/wundergraph/commit/bb3b6bd31250b402fe0c9a099b0dad993976cf39)) (@fiam)
* subscriptions, ws connection init hook ([#243](https://github.com/wundergraph/wundergraph/issues/243)) ([2e4ae85](https://github.com/wundergraph/wundergraph/commit/2e4ae8506558165a9bf3ada4b8f45cee55a9f18d)) (@YuriBuerov)

## [0.96.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.95.0...@wundergraph/protobuf@0.96.0) (2022-09-29)

### Features

* sse subscriptions configuration ([#217](https://github.com/wundergraph/wundergraph/issues/217)) ([7bf72ef](https://github.com/wundergraph/wundergraph/commit/7bf72efd16a8bac422db32fe957e102395d7357c)) (@YuriBuerov)

### Bug Fixes

* allow to configure internal and public node urls separately ([#207](https://github.com/wundergraph/wundergraph/issues/207)) ([c01bd9a](https://github.com/wundergraph/wundergraph/commit/c01bd9a1cedefbb5fd0ecde83f3b96b3dfee6f41)) (@spetrunin)

## [0.95.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.94.0...@wundergraph/protobuf@0.95.0) (2022-09-21)

### Features

* add post logout hook ([#196](https://github.com/wundergraph/wundergraph/issues/196)) ([6870130](https://github.com/wundergraph/wundergraph/commit/6870130b0c4e6fc269d81160994384c1d1cf6e59)) (@jensneuse)

## [0.94.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.93.0...@wundergraph/protobuf@0.94.0) (2022-09-19)

### Features

* implement config first approach ([#151](https://github.com/wundergraph/wundergraph/issues/151)) ([803da0e](https://github.com/wundergraph/wundergraph/commit/803da0e51beb3a7b23ee826dfde835eccfa1c2dd)), closes [#190](https://github.com/wundergraph/wundergraph/issues/190) (@spetrunin)

## [0.93.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.92.0...@wundergraph/protobuf@0.93.0) (2022-09-14)

### Features

* oidc,  allow passing additional query parameters to the IDP ([#178](https://github.com/wundergraph/wundergraph/issues/178)) ([d015bb1](https://github.com/wundergraph/wundergraph/commit/d015bb150762cba7a46865e66f3de633e731de07)) (@YuriBuerov)

## [0.92.0](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.91.3...@wundergraph/protobuf@0.92.0) (2022-09-01)

### Features

* implement userId for fromClaim directive ([#152](https://github.com/wundergraph/wundergraph/issues/152)) ([51df6e5](https://github.com/wundergraph/wundergraph/commit/51df6e50244bee9f5f8d579ff6f604e1a1c853d9)) (@jensneuse)

## [0.91.3](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.91.2...@wundergraph/protobuf@0.91.3) (2022-08-29)

**Note:** Version bump only for package @wundergraph/protobuf

## [0.91.2](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.91.1...@wundergraph/protobuf@0.91.2) (2022-08-18)

**Note:** Version bump only for package @wundergraph/protobuf

## [0.91.1](https://github.com/wundergraph/wundergraph/compare/@wundergraph/protobuf@0.91.0...@wundergraph/protobuf@0.91.1) (2022-06-02)

**Note:** Version bump only for package @wundergraph/protobuf

## 0.91.0

### Minor Changes

- [#18](https://github.com/wundergraph/wundergraph/pull/18) [`afea237`](https://github.com/wundergraph/wundergraph/commit/afea23771191e049aab5ce56ce775775389e8770) Thanks [@StarpTech](https://github.com/StarpTech)! - move wunderctl / go binary into local node_modules

## 0.90.1

### Patch Changes

- [`0857db3`](https://github.com/wundergraph/wundergraph/commit/0857db3d55209fb878fe6326629b125c6f2d2315) Thanks [@StarpTech](https://github.com/StarpTech)! - publish dist folder
