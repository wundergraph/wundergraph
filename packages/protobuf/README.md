# Protobuf

![wunderctl](https://img.shields.io/npm/v/@wundergraph/protobuf.svg)

> **Warning**: This package is not intended for standalone use. No api compatibility guarantees.
