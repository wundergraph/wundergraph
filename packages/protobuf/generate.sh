#!/usr/bin/env sh

PROTOC_GEN_TS_PROTO_PATH="./node_modules/.bin/protoc-gen-ts_proto"
PROTOC_GEN_TS_PATH="./node_modules/.bin/protoc-gen-ts"

OUT_DIR="./wgpb"

mkdir -p ${OUT_DIR}

protoc  \
  --plugin="${PROTOC_GEN_TS_PROTO_PATH}" \
  --proto_path=../../types/protos \
  --experimental_allow_proto3_optional \
  --ts_proto_opt=unrecognizedEnum=false \
  --ts_proto_opt=esModuleInterop=true \
  --ts_proto_opt=env=node \
  --ts_proto_opt=outputEncodeMethods=true \
  --ts_proto_opt=outputJsonMethods=true \
  --ts_proto_opt=outputClientImpl=false \
  --ts_proto_out="${OUT_DIR}" \
  ../../types/protos/*.proto
