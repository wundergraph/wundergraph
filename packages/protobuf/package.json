{
  "name": "@wundergraph/protobuf",
  "version": "0.116.0",
  "license": "Apache-2.0",
  "main": "./dist/wundernode_config.js",
  "types": "./dist/wundernode_config.d.ts",
  "description": "WunderGraph protobuf descriptions",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "registry": "https://registry.npmjs.org",
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/wundergraph/wundergraph.git"
  },
  "homepage": "https://wundergraph.com",
  "author": {
    "name": "WunderGraph Maintainers",
    "email": "info@wundergraph.com"
  },
  "bugs": {
    "url": "https://github.com/wundergraph/wundergraph/issues"
  },
  "scripts": {
    "build": "./generate.sh && tsc"
  },
  "dependencies": {
    "long": "^5.2.0",
    "protobufjs": "^6.11.2",
    "ts-proto": "^1.112.1"
  },
  "devDependencies": {
    "cpy-cli": "^4.1.0",
    "typescript": "^4.8.2"
  },
  "sideEffects": false
}
