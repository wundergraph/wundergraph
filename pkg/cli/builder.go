package cli

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/charmbracelet/bubbles/help"
	"github.com/charmbracelet/bubbles/key"
	"github.com/charmbracelet/bubbles/spinner"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

type ServerConfigLoaded struct {
	ServerURL                string
	Webhooks                 int32
	Operations               int32
	DatasourceConfigurations int32
	WunderctlVersion         string
	SdkVersion               string
	Authentication           bool
	FileUploads              bool
	PlaygroundEnabled        bool
}

type Error struct {
	Err error
}

type TaskStarted struct {
	Label string
}

type TaskEnded struct {
	Label string
	Err   error
}

type internalTask struct {
	Label     string
	StartTime time.Time
	Done      bool
}

type model struct {
	keys         keyMap
	help         help.Model
	err          error
	quitting     bool
	spinner      spinner.Model
	internalTask *internalTask
	serverConfig *ServerConfigLoaded
	viewHeight   int
	viewWidth    int
	ready        bool
}

func (m model) Init() tea.Cmd {
	return tea.Batch(
		m.spinner.Tick,
		tea.EnterAltScreen,
	)
}

type serverURLOpenFinished struct{ err error }

func OpenURL(serverURL string) tea.Cmd {
	return tea.ExecProcess(open(serverURL), func(err error) tea.Msg {
		return serverURLOpenFinished{err}
	})
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var (
		cmd  tea.Cmd
		cmds []tea.Cmd
	)

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch {
		case key.Matches(msg, m.keys.ClearConsole):
			return m, tea.ClearScreen
		case key.Matches(msg, m.keys.OpenBrowser):
			if m.serverConfig.ServerURL != "" {
				return m, OpenURL(m.serverConfig.ServerURL)
			}
		case key.Matches(msg, m.keys.Quit):
			m.quitting = true
			return m, tea.Quit
		case key.Matches(msg, m.keys.Documentation):
			return m, OpenURL("https://docs.wundergraph.com")
		case key.Matches(msg, m.keys.OpenIssue):
			return m, OpenURL("https://github.com/wundergraph/wundergraph/issues/new/choose")
		case key.Matches(msg, m.keys.Discord):
			return m, OpenURL("https://discord.com/invite/Jjmc8TC")
		}

	case tea.WindowSizeMsg:
		if !m.ready {
			m.viewHeight = msg.Height
			m.viewWidth = msg.Width
			// If we set a width on the help menu it can gracefully truncate
			// its view as needed.
			m.help.Width = m.viewWidth
			m.ready = true
		}
	case TaskStarted:
		m.internalTask = &internalTask{
			Label: msg.Label,
			Done:  false,
		}
		m.err = nil
	case TaskEnded:
		m.internalTask.Done = true
		m.internalTask.Label = msg.Label
		m.err = msg.Err
	case ServerConfigLoaded:
		m.serverConfig = &msg
	case Error:
		m.err = msg.Err
	case serverURLOpenFinished:
		m.err = msg.err
		if msg.err != nil {
			m.err = msg.err
			return m, tea.Quit
		}
	case spinner.TickMsg:
		m.spinner, cmd = m.spinner.Update(msg)
		return m, cmd
	}

	return m, tea.Batch(cmds...)
}

func (m model) View() string {
	if !m.ready {
		return "\n  Initializing..."
	}
	if m.quitting {
		return "\n  Quitting..."
	}

	docStyle := lipgloss.NewStyle().Padding(1, 2, 1, 2)

	if m.err != nil {
		doc := strings.Builder{}
		doc.WriteString(fmt.Sprintf("%s %s\n", FailStyle.Render("FAIL"), m.err.Error()))
		return docStyle.Render(doc.String())
	}

	doc := strings.Builder{}

	title := HeadlineTextStyle.Render("WunderGraph")
	sdkVersion := ""
	if m.serverConfig.SdkVersion != "" {
		sdkVersion = fmt.Sprintf(", SDK: %s", m.serverConfig.SdkVersion)
	}
	versionString := FeintStyle.Render(fmt.Sprintf("(CLI: %s%s)", m.serverConfig.WunderctlVersion, sdkVersion))

	doc.WriteString(fmt.Sprintf("%s %s\n\n", title, versionString))

	if m.serverConfig.ServerURL != "" {
		serverURL := lipgloss.JoinHorizontal(lipgloss.Left,
			BoldStyle.Render("➜ Local: "),
			FeintStyle.Render(m.serverConfig.ServerURL),
		)
		doc.WriteString(fmt.Sprintf("%s\n", serverURL))

		if m.serverConfig.PlaygroundEnabled {
			graphiQL := lipgloss.JoinHorizontal(lipgloss.Left,
				BoldStyle.Render("➜ Playground: "),
				fmt.Sprintf("%s/graphql", m.serverConfig.ServerURL),
			)
			doc.WriteString(fmt.Sprintf("%s\n", graphiQL))
		}

		logModeHint := lipgloss.JoinHorizontal(lipgloss.Left,
			BoldStyle.Render("➜ Logs:"),
			FeintStyle.Render(" use "),
			BoldStyle.Bold(true).Render("--logs"),
			FeintStyle.Render(" to enter log mode"),
		)
		doc.WriteString(fmt.Sprintf("%s\n\n", logModeHint))
	}

	if m.internalTask != nil {
		if m.internalTask.Done {
			doc.WriteString(fmt.Sprintf("%s %s\n", PassStyle.Render("SUCCESS"), m.internalTask.Label))
		} else {
			doc.WriteString(fmt.Sprintf("%s %s\n", m.spinner.View(), m.internalTask.Label))
		}
	}

	doc.WriteString("\n")

	if m.serverConfig != nil {
		authText := DisabledStyle.Render("✖ Auth")
		if m.serverConfig.Authentication {
			authText = TextStyle.Render("✔ Auth")
		}

		fileUploadText := DisabledStyle.Render("✖ File Uploads")
		if m.serverConfig.FileUploads {
			fileUploadText = TextStyle.Render("✔ File Uploads")
		}

		doc.WriteString(
			lipgloss.JoinHorizontal(lipgloss.Left, fmt.Sprintf("%s\n%s\n%s\n",
				lipgloss.NewStyle().Foreground(Purple100).PaddingRight(2).Render(fmt.Sprintf("%d DataSources", m.serverConfig.DatasourceConfigurations)),
				lipgloss.NewStyle().Foreground(Purple300).PaddingRight(2).Render(fmt.Sprintf("%d Operations", m.serverConfig.Operations)),
				lipgloss.NewStyle().Foreground(Purple600).PaddingRight(2).Render(fmt.Sprintf("%d Webhooks", m.serverConfig.Webhooks)),
			), fmt.Sprintf("%s\n%s\n",
				authText,
				fileUploadText,
			)),
		)

		doc.WriteString("\n")
	}

	helpView := m.help.View(m.keys)

	doc.WriteString(helpView + "\n")

	return docStyle.Render(doc.String())

}

type Options struct {
	ServerVersion string
}

func NewModel(ctx context.Context, options *Options) *tea.Program {
	s := spinner.New()
	s.Spinner = spinner.MiniDot
	s.Style = SpinnerStyle

	h := help.New()
	h.ShowAll = true
	h.Styles.ShortDesc = HelpStyle
	h.Styles.FullDesc = HelpStyle
	h.Styles.FullKey = TextStyle
	h.Styles.ShortKey = TextStyle
	h.Styles.ShortSeparator = SeparatorStyle

	model := model{
		keys:         keys,
		help:         h,
		spinner:      s,
		internalTask: &internalTask{},
		serverConfig: &ServerConfigLoaded{
			WunderctlVersion: options.ServerVersion,
		},
	}

	p := tea.NewProgram(
		model,
		tea.WithContext(ctx),
		tea.WithAltScreen(),
	)

	return p
}
