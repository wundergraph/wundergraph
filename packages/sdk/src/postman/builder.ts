import { OperationType } from '@wundergraph/protobuf';
import { JSONSchema7Definition } from 'json-schema';
import { Collection, Request } from 'postman-collection';
import { GraphQLOperation } from '../graphql/operations';

export interface PostmanBuilderOptions {
	baseURL: string;
}

export interface JSONSchemaParameterPath {
	path: string[];
	required: boolean;
	type: string;
}

// TS types of these modules are so bad, I opt out!
// docs: https://www.postmanlabs.com/postman-collection/

export const PostmanBuilder = (operations: GraphQLOperation[], options: PostmanBuilderOptions) => {
	const queryGroup = new Collection();
	queryGroup.id = 'Queries';
	queryGroup.name = 'Queries';
	queryGroup.describe('All your query operations');

	const mutationGroup = new Collection();
	mutationGroup.id = 'Mutations';
	mutationGroup.name = 'Mutations';
	mutationGroup.describe('All your mutation operations');

	operations.forEach((op) => {
		const operationURL = `{{apiBaseUrl}}/operations/${op.PathName}`;

		let paths: JSONSchemaParameterPath[] = [];
		buildPath([], false, op.VariablesSchema, paths);

		if (op.OperationType !== OperationType.MUTATION) {
			const request = queryRequestJson(operationURL, paths);

			queryGroup.items.add({
				id: op.Name,
				name: op.Name,
				request: request,
			});
		} else if (op.OperationType === OperationType.MUTATION) {
			const request = mutationRequestJson(operationURL, paths);

			mutationGroup.items.add({
				id: op.Name,
				name: op.Name,
				request: request,
			});
		}
	});

	const myCollection = new Collection();
	myCollection.id = 'Wundergraph';
	myCollection.describe('Your Wundergraph collection');
	myCollection.name = 'Wundergraph';

	// add variables
	myCollection.variables.add({
		key: 'apiBaseUrl',
		value: options.baseURL,
		type: 'string',
	});

	// add sub collections
	myCollection.items.add(queryGroup.toJSON());
	myCollection.items.add(mutationGroup.toJSON());

	return myCollection;
};

const mutationRequestJson = (url: string, paths: JSONSchemaParameterPath[]): string => {
	const request = new Request({
		url: url,
		method: 'POST',
		body: {
			mode: 'urlencoded',
			urlencoded: [],
			description: 'Your GraphQL variables in JSON',
		},
	});

	for (const path of paths) {
		request.body.urlencoded.add({
			key: path.path.join('.'),
			disabled: !path.required,
			description: `Type ${path.type}, ${path.required ? 'Required' : 'Optional'}`,
			value: '',
			type: 'text',
		});
	}

	return request.toJSON();
};

const queryRequestJson = (url: string, paths: JSONSchemaParameterPath[]): string => {
	const request = new Request({
		url: url,
		method: 'GET',
	});
	request.addHeader({
		key: 'Content-Type',
		value: 'application/json',
	});

	for (const path of paths) {
		request.addQueryParams([
			{
				key: path.path.join('.'),
				disabled: !path.required,
				value: '',
				description: `Type ${path.type}, ${path.required ? 'Required' : 'Optional'}`,
			},
		]);
	}

	return request.toJSON();
};

// TODO: Add "default" values
// path syntax follows https://github.com/tidwall/sjson#path-syntax
export function buildPath(
	path: string[],
	required: boolean,
	obj: JSONSchema7Definition,
	paths: JSONSchemaParameterPath[]
) {
	if (typeof obj === 'boolean') return;

	if (obj.type === 'array') {
		if (typeof obj.items === 'boolean') return;
		if (!obj.items) return;

		let items = [];
		if (!Array.isArray(obj.items)) items.push(obj.items);
		items.forEach((obj, index) => buildPath([...path, index.toString()], false, obj, paths));
		return;
	}

	// those nodes are limited due to recursion
	if (!obj.type) return;

	if (obj.properties) {
		Object.keys(obj.properties).forEach((key) => {
			if (obj.properties) {
				buildPath([...path, key], obj.required?.includes(key) || required, obj.properties[key], paths);
			}
		});
		return;
	}

	paths.push({
		path,
		required: false, // ignore it for now because this would make all variants required
		type: typeof obj.type === 'string' ? obj.type : 'any',
	});

	paths.sort(function (a, b) {
		return a.path.length - b.path.length;
	});
}
