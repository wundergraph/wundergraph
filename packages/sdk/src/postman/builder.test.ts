import { OperationExecutionEngine, OperationType } from '@wundergraph/protobuf';
import { Collection } from 'postman-collection';
import { PostmanBuilder } from './builder';

const operations = [
	{
		Name: 'GetRootObj',
		PathName: 'GetRootObj',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetRootObj2',
		PathName: 'GetRootObj2',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetUser',
		PathName: 'item/GetID',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetUser',
		PathName: 'location/GetID',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetUser',
		PathName: 'location/GetUser',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetUser',
		PathName: 'location/query/GetUser',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetCart',
		PathName: 'location/query/GetUser',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'GetSetUser',
		PathName: 'location/query/SetUser',
		OperationType: OperationType.MUTATION,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
	{
		Name: 'NewGetSetUser',
		PathName: 'location/query/new/SetUser',
		OperationType: OperationType.MUTATION,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: {
			type: 'object',
			properties: {},
			additionalProperties: false,
			definitions: {},
		},
	},
] as any;

test('PostmanBuilder', async () => {
	const result = PostmanBuilder(operations, { baseURL: 'http://localhost:9991' });

	expect(result).toMatchSnapshot();
});
