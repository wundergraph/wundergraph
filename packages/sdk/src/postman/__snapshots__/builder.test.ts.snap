// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostmanBuilder 1`] = `
{
  "event": [],
  "info": {
    "_postman_id": "Wundergraph",
    "description": {
      "content": "Your Wundergraph collection",
      "type": "text/plain",
    },
    "name": "Wundergraph",
    "schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
    "version": undefined,
  },
  "item": [
    {
      "_": {
        "postman_id": "operatations",
      },
      "event": [],
      "id": "operatations",
      "item": [
        {
          "event": [],
          "id": "GetRootObj",
          "name": "GetRootObj",
          "request": {
            "header": [
              {
                "key": "Content-Type",
                "value": "application/json",
              },
            ],
            "method": "GET",
            "url": {
              "host": [
                "{{apiBaseUrl}}",
              ],
              "path": [
                "operations",
                "GetRootObj",
              ],
              "query": [],
              "variable": [],
            },
          },
          "response": [],
        },
        {
          "event": [],
          "id": "GetRootObj2",
          "name": "GetRootObj2",
          "request": {
            "header": [
              {
                "key": "Content-Type",
                "value": "application/json",
              },
            ],
            "method": "GET",
            "url": {
              "host": [
                "{{apiBaseUrl}}",
              ],
              "path": [
                "operations",
                "GetRootObj2",
              ],
              "query": [],
              "variable": [],
            },
          },
          "response": [],
        },
        {
          "event": [],
          "id": "item",
          "item": [
            {
              "event": [],
              "id": "item/GetID",
              "name": "GetID",
              "request": {
                "header": [
                  {
                    "key": "Content-Type",
                    "value": "application/json",
                  },
                ],
                "method": "GET",
                "url": {
                  "host": [
                    "{{apiBaseUrl}}",
                  ],
                  "path": [
                    "operations",
                    "item",
                    "GetID",
                  ],
                  "query": [],
                  "variable": [],
                },
              },
              "response": [],
            },
          ],
          "name": "item",
        },
        {
          "event": [],
          "id": "location",
          "item": [
            {
              "event": [],
              "id": "location/GetID",
              "name": "GetID",
              "request": {
                "header": [
                  {
                    "key": "Content-Type",
                    "value": "application/json",
                  },
                ],
                "method": "GET",
                "url": {
                  "host": [
                    "{{apiBaseUrl}}",
                  ],
                  "path": [
                    "operations",
                    "location",
                    "GetID",
                  ],
                  "query": [],
                  "variable": [],
                },
              },
              "response": [],
            },
            {
              "event": [],
              "id": "location/GetUser",
              "name": "GetUser",
              "request": {
                "header": [
                  {
                    "key": "Content-Type",
                    "value": "application/json",
                  },
                ],
                "method": "GET",
                "url": {
                  "host": [
                    "{{apiBaseUrl}}",
                  ],
                  "path": [
                    "operations",
                    "location",
                    "GetUser",
                  ],
                  "query": [],
                  "variable": [],
                },
              },
              "response": [],
            },
            {
              "event": [],
              "id": "location/query",
              "item": [
                {
                  "event": [],
                  "id": "location/query/GetUser",
                  "name": "GetUser",
                  "request": {
                    "header": [
                      {
                        "key": "Content-Type",
                        "value": "application/json",
                      },
                    ],
                    "method": "GET",
                    "url": {
                      "host": [
                        "{{apiBaseUrl}}",
                      ],
                      "path": [
                        "operations",
                        "location",
                        "query",
                        "GetUser",
                      ],
                      "query": [],
                      "variable": [],
                    },
                  },
                  "response": [],
                },
                {
                  "event": [],
                  "id": "location/query/GetUser",
                  "name": "GetUser",
                  "request": {
                    "header": [
                      {
                        "key": "Content-Type",
                        "value": "application/json",
                      },
                    ],
                    "method": "GET",
                    "url": {
                      "host": [
                        "{{apiBaseUrl}}",
                      ],
                      "path": [
                        "operations",
                        "location",
                        "query",
                        "GetUser",
                      ],
                      "query": [],
                      "variable": [],
                    },
                  },
                  "response": [],
                },
                {
                  "event": [],
                  "id": "GetSetUser",
                  "name": "SetUser",
                  "request": {
                    "body": {
                      "mode": "urlencoded",
                      "urlencoded": [],
                    },
                    "method": "POST",
                    "url": {
                      "host": [
                        "{{apiBaseUrl}}",
                      ],
                      "path": [
                        "operations",
                        "location",
                        "query",
                        "SetUser",
                      ],
                      "query": [],
                      "variable": [],
                    },
                  },
                  "response": [],
                },
              ],
              "name": "query",
            },
            {
              "event": [],
              "id": "location/query",
              "name": "query",
              "request": {
                "method": "GET",
                "url": {
                  "query": [],
                  "variable": [],
                },
              },
              "response": [],
            },
          ],
          "name": "location",
        },
      ],
      "name": "operations",
    },
  ],
  "variable": [
    {
      "key": "apiBaseUrl",
      "type": "string",
      "value": "http://localhost:9991",
    },
  ],
}
`;
